### DISCUSSION TITLE
GCG: optimistic update throws type error when returning partial data

### DISCUSSION THREAD

**Discussion by drj17** [urql-graphql/urql#2320] [Q&A] [answered]
I'm trying to implement an optimistic update, updating the the phase of a projectInfluencer.  In this schema, `influencer` is a non nullable field on `projectInfluencer`, but my mutation is not returning the `influencer` field because it isn't necessary.

```gql
mutation UpdateProjectInfluencerPhase($params: UpdateProjectInfluencerInput!) {
  updateProjectInfluencer(params: $params) {
    projectInfluencer {
      id
      projectPhase {
        id
      }
    }
  }
}

```
```ts
const optimistic: GraphCacheOptimisticUpdaters = {
  updateProjectInfluencer: (variables, _cache, _info) => ({
    __typename: 'UpdateProjectInfluencerPayload',
    projectInfluencer: {
      __typename: 'ProjectInfluencer',
      id: variables.params.id,
      projectPhase: {
        __typename: 'ProjectPhase',
        id: variables.params.projectPhaseId,
      },
    },
  }),
};
```
 
I think because the `GraphCacheOptimisticUpdaters` uses the regular mutation payload which in turn uses the types generates from the schema, the partial data I'm returning doesn't satisfy the type.  The README says you only need to return the data that the mutation would be returning, so this seems like a bug with the generated types?  Or maybe I'm just missing something 😄 

![Screen Shot 2022-02-28 at 5 52 42 PM](https://user-images.githubusercontent.com/13426126/156071780-b919dbfb-f1d6-490a-a1ac-bb15ab96fe70.png)

urql version & exchanges:
```json
{
    "@graphql-codegen/typescript-urql-graphcache": "^2.2.6",
    "@urql/exchange-graphcache": "^4.3.6",
    "urql": "^2.1.3",
}

```

Let me know if this should be raised in the graphql-codegen repo instead

**Comment by JoviDeCroock** [comment, marked as answer]
Isn't the top typename a Mutation? I don't think it's mandatory but that seems a bit off. What is the exception you are getting? Edit sorry I thought with throw you meant a js err

Yes the types could be off for that one, I think it will be quite hard to remedy this one

**Reply by drj17** [reply]
It seems to work without the top typename, that's also how it's shown in the documentation.  I patched my package locally to recursively make the fields optional and nullable, which at least in my use case works well enough.  I'm not versed enough in typescript to say if this will negatively affect type safety but at least my code compiles now 😄

**Comment by ghost** [comment]
I encountered the same issue. The problems seems to be that `WithTypename` marks an entity as Partial but not its child entities. As @drj17 mentioned a solution would be making `WithTypename` recursive:
```
export type WithTypename<T> = T extends { __typename?: any }
  ? Partial<{
      [K in keyof T]: T[K] extends Array<infer S>
        ? Array<WithTypename<S>>
        : T[K] extends object | undefined
        ? WithTypename<T[K]>
        : T[K];
    }> & { __typename: NonNullable<T["__typename"]> }
  : T;
```
Is there any reason you would not want the child entities to be partial?
